File: android/app/src/main/kotlin/com/timecoin/wallet/network/MasternodeClient.kt
package com.timecoin.wallet.network

import android.util.Log
import com.timecoin.wallet.crypto.NetworkType
import com.timecoin.wallet.model.*
import io.ktor.client.*
import io.ktor.client.call.*
import io.ktor.client.engine.okhttp.*
import io.ktor.client.plugins.contentnegotiation.*
import io.ktor.client.request.*
import io.ktor.http.*
import io.ktor.serialization.kotlinx.json.*
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.*
import okhttp3.OkHttpClient
import java.security.SecureRandom
import java.security.cert.X509Certificate
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicLong
import javax.net.ssl.SSLContext
import javax.net.ssl.TrustManager
import javax.net.ssl.X509TrustManager

/**
 * Masternode JSON-RPC 2.0 client — mirrors the desktop wallet's
 * masternode_client.rs. Communicates with masternodes over HTTPS.
 * Masternodes use self-signed TLS certificates by default.
 *
 * Ports: 24001 (mainnet), 24101 (testnet)
 */
class MasternodeClient(
    endpoint: String,
    private val credentials: Pair<String, String>? = null,
) {
    val rpcEndpoint: String = if (endpoint.startsWith("http://") || endpoint.startsWith("https://")) {
        endpoint
    } else {
        "https://$endpoint"
    }

    private val client = HttpClient(OkHttp) {
        install(ContentNegotiation) { json(Json { ignoreUnknownKeys = true }) }
        engine {
            preconfigured = trustAllOkHttpClient
        }
    }

    private val requestId = AtomicLong(1)

    // ── RPC methods ──

    /** Get balance for a single address. */
    suspend fun getBalance(address: String): Balance {
        val result = rpcCall("getbalance", buildJsonArray { add(address) })
        val confirmed = jsonToSatoshis(result.jsonObject["available"])
        val total = jsonToSatoshis(result.jsonObject["balance"])
        return Balance(confirmed = confirmed, pending = total - confirmed, total = total)
    }

    /** Get combined balance across multiple addresses (batched to avoid oversized responses). */
    suspend fun getBalances(addresses: List<String>): Balance {
        if (addresses.size <= BATCH_SIZE) {
            val result = rpcCall("getbalances", buildJsonArray { add(buildJsonArray { addresses.forEach { add(it) } }) })
            val confirmed = jsonToSatoshis(result.jsonObject["available"])
            val total = jsonToSatoshis(result.jsonObject["balance"])
            return Balance(confirmed = confirmed, pending = total - confirmed, total = total)
        }
        // Batch: sum balances across chunks
        var confirmed = 0L
        var total = 0L
        for (chunk in addresses.chunked(BATCH_SIZE)) {
            val result = rpcCall("getbalances", buildJsonArray { add(buildJsonArray { chunk.forEach { add(it) } }) })
            confirmed += jsonToSatoshis(result.jsonObject["available"])
            total += jsonToSatoshis(result.jsonObject["balance"])
        }
        return Balance(confirmed = confirmed, pending = total - confirmed, total = total)
    }

    /** Get transaction history for a single address. */
    suspend fun getTransactions(address: String, limit: Int = 50): List<TransactionRecord> {
        val result = rpcCall("listtransactions", buildJsonArray { add(address); add(limit) })
        return parseTransactionList(result)
    }

    /** Get transaction history across multiple addresses (batched). */
    suspend fun getTransactionsMulti(addresses: List<String>, limit: Int = 50): List<TransactionRecord> {
        if (addresses.size <= BATCH_SIZE) {
            val result = rpcCall("listtransactionsmulti", buildJsonArray {
                add(buildJsonArray { addresses.forEach { add(it) } }); add(limit)
            })
            return parseTransactionList(result)
        }
        // Batch: collect and deduplicate by uniqueKey
        val all = mutableListOf<TransactionRecord>()
        for (chunk in addresses.chunked(BATCH_SIZE)) {
            val result = rpcCall("listtransactionsmulti", buildJsonArray {
                add(buildJsonArray { chunk.forEach { add(it) } }); add(limit)
            })
            all.addAll(parseTransactionList(result))
        }
        val seen = mutableSetOf<String>()
        return all.filter { seen.add(it.uniqueKey) }
    }

    /** Get UTXOs for addresses (batched). */
    suspend fun getUtxos(addresses: List<String>): List<Utxo> {
        val chunks = if (addresses.size <= BATCH_SIZE) listOf(addresses) else addresses.chunked(BATCH_SIZE)
        val all = mutableListOf<Utxo>()
        for (chunk in chunks) {
            val result = rpcCall("listunspentmulti", buildJsonArray {
                add(buildJsonArray { chunk.forEach { add(it) } })
            })
            all.addAll(result.jsonArray.mapNotNull { u ->
                val obj = u.jsonObject
                val txid = obj["txid"]?.jsonPrimitive?.contentOrNull ?: return@mapNotNull null
                val vout = obj["vout"]?.jsonPrimitive?.intOrNull ?: return@mapNotNull null
                val amount = jsonToSatoshis(obj["amount"])
                val addr = obj["address"]?.jsonPrimitive?.contentOrNull ?: ""
                val confirmations = obj["confirmations"]?.jsonPrimitive?.intOrNull ?: 0
                val spendable = obj["spendable"]?.jsonPrimitive?.booleanOrNull ?: true
                Utxo(txid, vout, amount, addr, confirmations, spendable)
            })
        }
        return all
    }

    /** Broadcast a signed transaction (hex-encoded). */
    suspend fun broadcastTransaction(txHex: String): String {
        val result = rpcCall("sendrawtransaction", buildJsonArray { add(txHex) })
        return result.jsonPrimitive.contentOrNull ?: result.toString().trim('"')
    }

    /** Validate an address on the masternode. */
    suspend fun validateAddress(address: String): Boolean {
        val result = rpcCall("validateaddress", buildJsonArray { add(address) })
        return result.jsonObject["isvalid"]?.jsonPrimitive?.booleanOrNull ?: false
    }

    /** Get blockchain info (health check). */
    suspend fun healthCheck(): HealthStatus {
        val result = rpcCall("getblockchaininfo", buildJsonArray {})
        val height = result.jsonObject["blocks"]?.jsonPrimitive?.longOrNull
            ?: result.jsonObject["height"]?.jsonPrimitive?.longOrNull ?: 0
        val network = result.jsonObject["chain"]?.jsonPrimitive?.contentOrNull ?: "unknown"

        var peerCount = 0
        var version = ""
        try {
            val ni = rpcCall("getnetworkinfo", buildJsonArray {})
            peerCount = ni.jsonObject["connections"]?.jsonPrimitive?.intOrNull ?: 0
            version = ni.jsonObject["subversion"]?.jsonPrimitive?.contentOrNull?.trim('/') ?: ""
        } catch (_: Exception) { }

        return HealthStatus(
            status = "healthy",
            version = "$network ($version)",
            blockHeight = height,
            peerCount = peerCount,
        )
    }

    /** Get current block height. */
    suspend fun getBlockHeight(): Long {
        val result = rpcCall("getblockcount", buildJsonArray {})
        return result.jsonPrimitive.longOrNull ?: 0
    }

    /** Query instant finality status. */
    suspend fun getTransactionFinality(txid: String): FinalityStatus {
        val result = rpcCall("gettransactionfinality", buildJsonArray { add(txid) })
        return FinalityStatus(
            txid = txid,
            finalized = result.jsonObject["finalized"]?.jsonPrimitive?.booleanOrNull ?: false,
            confirmations = result.jsonObject["confirmations"]?.jsonPrimitive?.intOrNull ?: 0,
        )
    }

    /**
     * Get peer list from the connected masternode (gossip discovery).
     * Returns a list of "IP:P2P_PORT" strings from the masternode's peer table.
     */
    suspend fun getPeerAddresses(): List<String> {
        val result = rpcCall("getpeerinfo", buildJsonArray {})
        return result.jsonArray.mapNotNull { peer ->
            peer.jsonObject["addr"]?.jsonPrimitive?.contentOrNull
        }
    }

    /**
     * Get full transaction details including per-output breakdown.
     * Used to expand "consolidate" (self-send) entries into individual outputs
     * so processTransactions can create proper send + receive + fee entries.
     */
    suspend fun getTransactionDetail(txid: String): Pair<Long, List<TxOutputInfo>> {
        val result = rpcCall("gettransaction", buildJsonArray { add(txid) })
        val obj = result.jsonObject
        val fee = jsonToSatoshisAbs(obj["fee"])
        val outputs = obj["vout"]?.jsonArray?.mapNotNull { vout ->
            val vo = vout.jsonObject
            val value = jsonToSatoshisAbs(vo["value"])
            val n = vo["n"]?.jsonPrimitive?.intOrNull ?: return@mapNotNull null
            val addr = vo["scriptPubKey"]?.jsonObject?.get("address")
                ?.jsonPrimitive?.contentOrNull ?: return@mapNotNull null
            TxOutputInfo(value = value, index = n, address = addr)
        } ?: emptyList()
        return Pair(fee, outputs)
    }

    /**
     * Send a payment request to another wallet via masternode relay.
     * The masternode forwards it as a `payment_request` WS event to the payer's subscribed address.
     */
    suspend fun sendPaymentRequest(
        requestId: String,
        requesterAddress: String,
        payerAddress: String,
        amountSats: Long,
        memo: String,
        requesterName: String,
    ): Boolean = try {
        rpcCall("sendpaymentrequest", buildJsonArray {
            add(buildJsonObject {
                put("id", requestId)
                put("requester_address", requesterAddress)
                put("payer_address", payerAddress)
                put("amount", amountSats)
                put("memo", memo)
                put("requester_name", requesterName)
            })
        })
        true
    } catch (e: Exception) {
        Log.w("MasternodeClient", "sendPaymentRequest failed: ${e.message}")
        false
    }

    /** Notify the masternode that the payer has viewed a payment request. Requester is notified via `payment_request_viewed`. */
    suspend fun markPaymentRequestViewed(requestId: String, payerAddress: String): Boolean = try {
        rpcCall("markpaymentrequestviewed", buildJsonArray {
            add(buildJsonObject {
                put("id", requestId)
                put("payer_address", payerAddress)
            })
        })
        true
    } catch (e: Exception) {
        Log.w("MasternodeClient", "markPaymentRequestViewed failed: ${e.message}")
        false
    }

    /** Cancel a previously sent payment request. Masternode notifies payer via `payment_request_cancelled`. */
    suspend fun cancelPaymentRequest(requestId: String, requesterAddress: String): Boolean = try {
        rpcCall("cancelpaymentrequest", buildJsonArray {
            add(buildJsonObject {
                put("id", requestId)
                put("requester_address", requesterAddress)
            })
        })
        true
    } catch (e: Exception) {
        Log.w("MasternodeClient", "cancelPaymentRequest failed: ${e.message}")
        false
    }

    /** Respond to a payment request. Masternode notifies requester via `payment_request_response`. */
    suspend fun respondToPaymentRequest(
        requestId: String,
        payerAddress: String,
        accepted: Boolean,
    ): Boolean = try {
        rpcCall("respondpaymentrequest", buildJsonArray {
            add(buildJsonObject {
                put("id", requestId)
                put("payer_address", payerAddress)
                put("accepted", accepted)
            })
        })
        true
    } catch (e: Exception) {
        Log.w("MasternodeClient", "respondToPaymentRequest failed: ${e.message}")
        false
    }

    fun close() { client.close() }

    // ── Internal ──

    private suspend fun rpcCall(method: String, params: JsonElement): JsonElement {
        val id = requestId.getAndIncrement()
        val request = buildJsonObject {
            put("jsonrpc", "2.0")
            put("id", id.toString())
            put("method", method)
            put("params", params)
        }

        val response: JsonObject = client.post(rpcEndpoint) {
            contentType(ContentType.Application.Json)
            if (credentials != null) {
                basicAuth(credentials.first, credentials.second)
            }
            setBody(request)
        }.body()

        val error = response["error"]
        if (error != null && error !is JsonNull) {
            val code = error.jsonObject["code"]?.jsonPrimitive?.longOrNull ?: -1
            val message = error.jsonObject["message"]?.jsonPrimitive?.contentOrNull ?: "Unknown error"
            throw MasternodeException("RPC error $code: $message")
        }

        return response["result"] ?: throw MasternodeException("No result in JSON-RPC response")
    }

    private fun parseTransactionList(result: JsonElement): List<TransactionRecord> {
        val txArray = if (result is JsonObject && result.containsKey("transactions")) {
            result["transactions"]?.jsonArray ?: return emptyList()
        } else {
            result.jsonArray
        }
        return txArray.mapNotNull { tx ->
            val obj = tx.jsonObject
            val txid = obj["txid"]?.jsonPrimitive?.contentOrNull ?: return@mapNotNull null
            val category = obj["category"]?.jsonPrimitive?.contentOrNull ?: "unknown"
            val rawAmount = jsonToSatoshisAbs(obj["amount"])
            val fee = jsonToSatoshisAbs(obj["fee"])
            val address = obj["address"]?.jsonPrimitive?.contentOrNull ?: ""
            val vout = obj["vout"]?.jsonPrimitive?.intOrNull ?: 0
            val timestamp = obj["time"]?.jsonPrimitive?.longOrNull
                ?: obj["blocktime"]?.jsonPrimitive?.longOrNull
                ?: obj["timestamp"]?.jsonPrimitive?.longOrNull ?: 0
            val inBlock = obj["blockhash"]?.jsonPrimitive?.contentOrNull != null
            val blockHash = obj["blockhash"]?.jsonPrimitive?.contentOrNull ?: ""
            val finalized = obj["finalized"]?.jsonPrimitive?.booleanOrNull ?: false
            val blockHeight = obj["blockheight"]?.jsonPrimitive?.longOrNull ?: 0
            val confirmations = obj["confirmations"]?.jsonPrimitive?.longOrNull ?: 0
            val txStatus = if (inBlock || finalized) TransactionStatus.Approved else TransactionStatus.Pending
            val rpcMemo = obj["memo"]?.jsonPrimitive?.contentOrNull ?: ""

            val isSend = category == "send"
            val isConsolidate = category == "consolidate"
            val isBlockReward = category == "generate"

            Log.d("MasternodeClient", "parseTx: txid=${txid.take(12)}.. cat=$category " +
                "rawAmt=$rawAmount fee=$fee addr=${address.take(16)}.. vout=$vout")

            // "consolidate" = self-send (all outputs go back to own addresses).
            // Pass through as a marker so refreshTransactions can expand it
            // via gettransaction into individual per-output entries.
            if (isConsolidate) {
                return@mapNotNull TransactionRecord(
                    txid = txid,
                    vout = vout,
                    isSend = true,
                    isConsolidate = true,
                    address = address,
                    amount = rawAmount,
                    fee = fee,
                    timestamp = timestamp,
                    status = txStatus,
                    blockHash = blockHash,
                    blockHeight = blockHeight,
                    confirmations = confirmations,
                )
            }

            // RPC includes fee in the send amount — subtract it so we
            // display only the actual transferred value. Matches desktop
            // wallet behavior (masternode_client.rs:311-312).
            val displayAmount = if (isSend && fee > 0) {
                maxOf(rawAmount - fee, 0)
            } else {
                rawAmount
            }

            // Skip zero-amount received entries — these are staking inputs with
            // no corresponding payout. Block rewards (generate) are kept even
            // at zero so they appear in history.
            if (displayAmount == 0L && !isSend && !isBlockReward) return@mapNotNull null

            TransactionRecord(
                txid = txid,
                vout = vout,
                isSend = isSend,
                address = address,
                amount = displayAmount,
                fee = fee,
                timestamp = timestamp,
                status = txStatus,
                blockHash = blockHash,
                blockHeight = blockHeight,
                confirmations = confirmations,
                memo = if (isBlockReward) "Block Reward" else rpcMemo,
            )
        }
    }

    companion object {
        /** Max addresses per RPC call to avoid oversized responses. */
        private const val BATCH_SIZE = 20

        /** Trust manager that accepts all certificates (for self-signed masternode certs). */
        private val trustAllManager = object : X509TrustManager {
            override fun checkClientTrusted(chain: Array<X509Certificate>?, authType: String?) {}
            override fun checkServerTrusted(chain: Array<X509Certificate>?, authType: String?) {}
            override fun getAcceptedIssuers(): Array<X509Certificate> = arrayOf()
        }

        private val trustAllSslContext: SSLContext = SSLContext.getInstance("TLS").apply {
            init(null, arrayOf<TrustManager>(trustAllManager), SecureRandom())
        }

        /** Pre-configured OkHttpClient that trusts all certs and forces HTTP/1.1. */
        val trustAllOkHttpClient: OkHttpClient = OkHttpClient.Builder()
            .connectTimeout(10, TimeUnit.SECONDS)
            .readTimeout(30, TimeUnit.SECONDS)
            .sslSocketFactory(trustAllSslContext.socketFactory, trustAllManager)
            .hostnameVerifier { _, _ -> true }
            .protocols(listOf(okhttp3.Protocol.HTTP_1_1))
            .build()

        /** Parse a JSON value to satoshis (1 TIME = 100_000_000 satoshis). */
        fun jsonToSatoshis(value: JsonElement?): Long {
            if (value == null || value is JsonNull) return 0
            val s = when (value) {
                is JsonPrimitive -> value.content
                else -> return 0
            }.trim().trimStart('-')
            // If original was negative, return 0
            if (value is JsonPrimitive && value.content.trim().startsWith('-')) return 0
            return parseTimeStringToSatoshis(s)
        }

        private fun jsonToSatoshisAbs(value: JsonElement?): Long {
            if (value == null || value is JsonNull) return 0
            val s = when (value) {
                is JsonPrimitive -> value.content
                else -> return 0
            }.trim().trimStart('-')
            return parseTimeStringToSatoshis(s)
        }

        private fun parseTimeStringToSatoshis(s: String): Long {
            if (s.contains('e', ignoreCase = true)) {
                return try { (s.toDouble() * 100_000_000).toLong() } catch (_: Exception) { 0 }
            }
            val parts = s.split('.', limit = 2)
            val whole = parts[0].toLongOrNull() ?: 0
            val frac = if (parts.size > 1) {
                parts[1].take(8).padEnd(8, '0').toLongOrNull() ?: 0
            } else 0
            return whole * 100_000_000 + frac
        }
    }
}

class MasternodeException(message: String) : Exception(message)

File: ios/TimeCoinWallet/Network/MasternodeClient.swift
import Foundation
import Alamofire

// JSON-RPC 2.0 client for masternode communication.
// Mirrors Android MasternodeClient: same RPC methods, same batching (20 per call),
// same trust-all TLS for self-signed masternode certs.
class MasternodeClient {
    let rpcEndpoint: String
    private let credentials: (String, String)?
    private let session: Session
    private var requestIdCounter: Int = 1

    static let batchSize = 20

    init(endpoint: String, credentials: (String, String)? = nil) {
        if endpoint.hasPrefix("http://") || endpoint.hasPrefix("https://") {
            self.rpcEndpoint = endpoint
        } else {
            self.rpcEndpoint = "https://\(endpoint)"
        }
        self.credentials = credentials

        // Trust all TLS certs (masternodes use self-signed)
        let trustManager = ServerTrustManager(allHostsMustBeEvaluated: false, evaluators: [:])
        let configuration = URLSessionConfiguration.default
        configuration.timeoutIntervalForRequest = 10
        configuration.timeoutIntervalForResource = 30
        self.session = Session(configuration: configuration, serverTrustManager: trustManager)
    }

    // MARK: - RPC methods

    func getBalance(address: String) async throws -> Balance {
        let result = try await rpcCall(method: "getbalance", params: [address])
        return parseBalance(result)
    }

    func getBalances(addresses: [String]) async throws -> Balance {
        if addresses.count <= Self.batchSize {
            let result = try await rpcCall(method: "getbalances", params: [addresses])
            return parseBalance(result)
        }
        var confirmed: Int64 = 0; var total: Int64 = 0; var locked: Int64 = 0
        for chunk in addresses.chunked(into: Self.batchSize) {
            let result = try await rpcCall(method: "getbalances", params: [chunk])
            let b = parseBalance(result)
            confirmed += b.confirmed; total += b.total; locked += b.locked
        }
        return Balance(confirmed: confirmed, pending: total - confirmed - locked, total: total, locked: locked)
    }

    func getTransactions(address: String, limit: Int = 50) async throws -> [TransactionRecord] {
        let result = try await rpcCall(method: "listtransactions", params: [address, limit])
        return parseTransactionList(result)
    }

    func getTransactionsMulti(addresses: [String], limit: Int = 50, offset: Int = 0) async throws -> [TransactionRecord] {
        if addresses.count <= Self.batchSize {
            let result = try await rpcCall(method: "listtransactionsmulti", params: [addresses, limit, offset])
            return parseTransactionList(result)
        }
        var all: [TransactionRecord] = []
        var seen = Set<String>()
        for chunk in addresses.chunked(into: Self.batchSize) {
            let result = try await rpcCall(method: "listtransactionsmulti", params: [chunk, limit, offset])
            for tx in parseTransactionList(result) {
                if seen.insert(tx.uniqueKey).inserted { all.append(tx) }
            }
        }
        return all
    }

    func getUtxos(addresses: [String]) async throws -> [Utxo] {
        var all: [Utxo] = []
        for chunk in (addresses.count <= Self.batchSize ? [addresses] : addresses.chunked(into: Self.batchSize)) {
            let result = try await rpcCall(method: "listunspentmulti", params: [chunk])
            if let arr = result as? [[String: Any]] {
                for u in arr {
                    guard let txid = u["txid"] as? String,
                          let vout = u["vout"] as? Int else { continue }
                    let amount = jsonToSatoshis(u["amount"])
                    let addr = u["address"] as? String ?? ""
                    let confs = u["confirmations"] as? Int ?? 0
                    let spendable = u["spendable"] as? Bool ?? true
                    all.append(Utxo(txid: txid, vout: vout, amount: amount, address: addr, confirmations: confs, spendable: spendable))
                }
            }
        }
        return all
    }

    func broadcastTransaction(_ tx: TimeCoinTransaction) async throws -> String {
        let encoder = JSONEncoder()
        let txData = try encoder.encode(tx)
        let txObj = try JSONSerialization.jsonObject(with: txData)
        let result = try await rpcCall(method: "sendrawtransaction", params: [txObj])
        return (result as? String) ?? "\(result)"
    }

    func validateAddress(_ address: String) async throws -> Bool {
        let result = try await rpcCall(method: "validateaddress", params: [address])
        if let obj = result as? [String: Any] {
            return obj["isvalid"] as? Bool ?? false
        }
        return false
    }

    func healthCheck() async throws -> HealthStatus {
        let result = try await rpcCall(method: "getblockchaininfo", params: [])
        guard let obj = result as? [String: Any] else {
            throw MasternodeError.unexpectedResponse
        }
        let height = obj["blocks"] as? Int64 ?? obj["height"] as? Int64 ?? 0
        let chain = obj["chain"] as? String ?? "unknown"
        let isSyncing = obj["initialblockdownload"] as? Bool ?? false
        let syncProgress = obj["verificationprogress"] as? Double ?? 1.0

        var peerCount = 0; var version = ""
        if let ni = try? await rpcCall(method: "getnetworkinfo", params: []),
           let niObj = ni as? [String: Any] {
            peerCount = niObj["connections"] as? Int ?? 0
            version = ((niObj["subversion"] as? String) ?? "").trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        }
        return HealthStatus(status: "healthy", version: "\(chain) (\(version))",
                           blockHeight: height, peerCount: peerCount,
                           isSyncing: isSyncing, syncProgress: syncProgress)
    }

    func getBlockHeight() async throws -> Int64 {
        let result = try await rpcCall(method: "getblockcount", params: [])
        return result as? Int64 ?? 0
    }

    func getGenesisHash() async throws -> String {
        let result = try await rpcCall(method: "getblockhash", params: [0])
        guard let hash = result as? String else { throw MasternodeError.unexpectedResponse }
        return hash
    }

    func getTransactionFinality(txid: String) async throws -> FinalityStatus {
        let result = try await rpcCall(method: "gettransactionfinality", params: [txid])
        guard let obj = result as? [String: Any] else { throw MasternodeError.unexpectedResponse }
        return FinalityStatus(
            txid: txid,
            finalized: obj["finalized"] as? Bool ?? false,
            confirmations: obj["confirmations"] as? Int ?? 0
        )
    }

    func getPeerAddresses() async throws -> [String] {
        let result = try await rpcCall(method: "getpeerinfo", params: [])
        guard let arr = result as? [[String: Any]] else { return [] }
        return arr.compactMap { $0["addr"] as? String }
    }

    func getTransactionDetail(txid: String) async throws -> TxDetail {
        let result = try await rpcCall(method: "gettransaction", params: [txid])
        guard let obj = result as? [String: Any] else { throw MasternodeError.unexpectedResponse }
        let fee = jsonToSatoshisAbs(obj["fee"])
        let finalized = obj["finalized"] as? Bool ?? false
        var outputs: [TxOutputInfo] = []
        if let vouts = obj["vout"] as? [[String: Any]] {
            for vout in vouts {
                let value = jsonToSatoshisAbs(vout["value"])
                guard let n = vout["n"] as? Int else { continue }
                let addr = (vout["scriptPubKey"] as? [String: Any])?["address"] as? String ?? ""
                outputs.append(TxOutputInfo(value: value, index: n, address: addr))
            }
        }
        return TxDetail(fee: fee, outputs: outputs, finalized: finalized)
    }

    func getBlockRewardBreakdown(height: Int64) async throws -> BlockRewardBreakdown {
        let result = try await rpcCall(method: "getblock", params: [height])
        guard let obj = result as? [String: Any] else { return BlockRewardBreakdown(blockHeight: height, rewards: []) }
        let rewards = (obj["masternode_rewards"] as? [[String: Any]] ?? []).compactMap { r -> BlockRewardEntry? in
            guard let addr = r["address"] as? String, let amount = r["amount"] as? Int64 else { return nil }
            return BlockRewardEntry(address: addr, amount: amount)
        }
        return BlockRewardBreakdown(blockHeight: height, rewards: rewards)
    }

    func getAddressPubkey(address: String) async throws -> String? {
        let result = try? await rpcCall(method: "getaddresspubkey", params: [address])
        return (result as? String).flatMap { $0.isEmpty ? nil : $0 }
    }

    func registerAddressPubkey(address: String, pubkeyHex: String) async throws -> Bool {
        _ = try? await rpcCall(method: "registeraddresspubkey", params: [address, pubkeyHex])
        return true
    }

    func sendPaymentRequest(
        requestId: String, requesterAddress: String, payerAddress: String,
        amountSats: Int64, memo: String, requesterName: String,
        pubkeyHex: String = "", signatureHex: String = "",
        timestamp: Int64 = Int64(Date().timeIntervalSince1970)
    ) async throws -> Bool {
        let payload: [String: Any] = [
            "id": requestId, "requester_address": requesterAddress,
            "payer_address": payerAddress, "amount": amountSats,
            "memo": memo, "requester_name": requesterName,
            "pubkey_hex": pubkeyHex, "signature_hex": signatureHex, "timestamp": timestamp
        ]
        _ = try? await rpcCall(method: "sendpaymentrequest", params: [payload])
        return true
    }

    func markPaymentRequestViewed(requestId: String, payerAddress: String) async throws -> Bool {
        let payload: [String: Any] = ["id": requestId, "payer_address": payerAddress]
        _ = try? await rpcCall(method: "markpaymentrequestviewed", params: [payload])
        return true
    }

    func cancelPaymentRequest(requestId: String, requesterAddress: String) async throws -> Bool {
        let payload: [String: Any] = ["id": requestId, "requester_address": requesterAddress]
        _ = try? await rpcCall(method: "cancelpaymentrequest", params: [payload])
        return true
    }

    func respondToPaymentRequest(requestId: String, payerAddress: String, accepted: Bool) async throws -> Bool {
        let payload: [String: Any] = ["id": requestId, "payer_address": payerAddress, "accepted": accepted]
        _ = try? await rpcCall(method: "respondpaymentrequest", params: [payload])
        return true
    }

    // MARK: - Internal

    private func rpcCall(method: String, params: [Any]) async throws -> Any {
        let id = requestIdCounter
        requestIdCounter += 1

        var body: [String: Any] = [
            "jsonrpc": "2.0",
            "id": "\(id)",
            "method": method,
            "params": params
        ]

        return try await withCheckedThrowingContinuation { continuation in
            var headers: HTTPHeaders = [.contentType("application/json")]
            if let (user, pass) = credentials {
                let cred = "\(user):\(pass)".data(using: .utf8)!.base64EncodedString()
                headers.add(.authorization("Basic \(cred)"))
            }

            session.request(rpcEndpoint, method: .post,
                           parameters: body,
                           encoding: JSONEncoding.default,
                           headers: headers)
            .validate()
            .responseJSON { response in
                switch response.result {
                case .success(let value):
                    guard let obj = value as? [String: Any] else {
                        continuation.resume(throwing: MasternodeError.unexpectedResponse)
                        return
                    }
                    if let error = obj["error"] as? [String: Any] {
                        let code = error["code"] as? Int ?? -1
                        let msg = error["message"] as? String ?? "Unknown error"
                        continuation.resume(throwing: MasternodeError.rpcError(code, msg))
                        return
                    }
                    guard let result = obj["result"] else {
                        continuation.resume(throwing: MasternodeError.noResult)
                        return
                    }
                    continuation.resume(returning: result)
                case .failure(let error):
                    continuation.resume(throwing: MasternodeError.networkError(error.localizedDescription))
                }
            }
        }
    }

    private func parseBalance(_ result: Any) -> Balance {
        guard let obj = result as? [String: Any] else { return .zero }
        let confirmed = jsonToSatoshis(obj["available"])
        let total = jsonToSatoshis(obj["balance"])
        let locked = jsonToSatoshis(obj["locked"])
        return Balance(confirmed: confirmed, pending: total - confirmed - locked, total: total, locked: locked)
    }

    private func parseTransactionList(_ result: Any) -> [TransactionRecord] {
        var arr: [[String: Any]] = []
        if let obj = result as? [String: Any], let txs = obj["transactions"] as? [[String: Any]] {
            arr = txs
        } else if let direct = result as? [[String: Any]] {
            arr = direct
        }

        return arr.compactMap { obj in
            guard let txid = obj["txid"] as? String else { return nil }
            let category = obj["category"] as? String ?? "unknown"
            let rawAmount = jsonToSatoshisAbs(obj["amount"])
            let fee = jsonToSatoshisAbs(obj["fee"])
            let address = obj["address"] as? String ?? ""
            let vout = obj["vout"] as? Int ?? 0
            let timestamp = obj["time"] as? Int64 ?? obj["blocktime"] as? Int64 ?? obj["timestamp"] as? Int64 ?? 0
            let inBlock = obj["blockhash"] as? String != nil
            let blockHash = obj["blockhash"] as? String ?? ""
            let finalized = obj["finalized"] as? Bool ?? false
            let blockHeight = obj["blockheight"] as? Int64 ?? 0
            let confirmations = obj["confirmations"] as? Int64 ?? 0
            let txStatus: TransactionStatus = (inBlock || finalized) ? .approved : .pending
            let rpcMemo = obj["memo"] as? String ?? ""

            let isSend = category == "send"
            let isConsolidate = category == "consolidate"
            let isBlockReward = category == "generate"

            if isConsolidate {
                return TransactionRecord(txid: txid, vout: vout, isSend: true, address: address,
                                         amount: rawAmount, fee: fee, timestamp: timestamp, status: txStatus,
                                         isConsolidate: true, blockHash: blockHash, blockHeight: blockHeight,
                                         confirmations: confirmations)
            }

            let displayAmount = isSend && fee > 0 ? max(rawAmount - fee, 0) : rawAmount
            if displayAmount == 0 && !isSend && !isBlockReward { return nil }

            return TransactionRecord(txid: txid, vout: vout, isSend: isSend, address: address,
                                     amount: displayAmount, fee: fee, timestamp: timestamp, status: txStatus,
                                     blockHash: blockHash, blockHeight: blockHeight, confirmations: confirmations,
                                     memo: isBlockReward ? "Block Reward" : rpcMemo)
        }
    }

    // MARK: - Amount parsing

    static func jsonToSatoshis(_ value: Any?) -> Int64 {
        guard let value = value else { return 0 }
        let s: String
        if let d = value as? Double { s = String(d) }
        else if let i = value as? Int64 { s = String(i) }
        else if let str = value as? String { s = str }
        else { return 0 }
        let trimmed = s.trimmingCharacters(in: .whitespaces)
        if trimmed.hasPrefix("-") { return 0 }
        return parseTimeString(trimmed)
    }

    private func jsonToSatoshis(_ value: Any?) -> Int64 { Self.jsonToSatoshis(value) }

    private func jsonToSatoshisAbs(_ value: Any?) -> Int64 {
        guard let value = value else { return 0 }
        let s: String
        if let d = value as? Double { s = String(abs(d)) }
        else if let i = value as? Int64 { s = String(abs(i)) }
        else if let str = value as? String { s = str.trimmingCharacters(in: .whitespaces).trimmingCharacters(in: CharacterSet(charactersIn: "-")) }
        else { return 0 }
        return Self.parseTimeString(s)
    }

    static func parseTimeString(_ s: String) -> Int64 {
        if s.lowercased().contains("e") {
            return Int64((Double(s) ?? 0) * 100_000_000)
        }
        let parts = s.components(separatedBy: ".")
        let whole = Int64(parts[0]) ?? 0
        let frac: Int64
        if parts.count > 1 {
            let fracStr = parts[1].prefix(8).padding(toLength: 8, withPad: "0", startingAt: 0)
            frac = Int64(fracStr) ?? 0
        } else {
            frac = 0
        }
        return whole * 100_000_000 + frac
    }
}

enum MasternodeError: Error, LocalizedError {
    case rpcError(Int, String)
    case noResult
    case unexpectedResponse
    case networkError(String)

    var errorDescription: String? {
        switch self {
        case .rpcError(let code, let msg): return "RPC error \(code): \(msg)"
        case .noResult: return "No result in JSON-RPC response"
        case .unexpectedResponse: return "Unexpected response format"
        case .networkError(let msg): return "Network error: \(msg)"
        }
    }
}

extension Array {
    func chunked(into size: Int) -> [[Element]] {
        stride(from: 0, to: count, by: size).map {
            Array(self[$0..<Swift.min($0 + size, count)])
        }
    }
}
